README.md
# babel-plugin-precompile-intl

This babel plugin compiles translations in the ICU message syntax into executable functions at build time,
analyzing the used features and importing the required runtime helpers from [precompile-intl-runtime](https://github.com/cibernox/precompile-intl-runtime)
to perform the operations that can't be performed at build time.

### Differences with other internationalization libraries

This approach is different than the taken by libraries like [intl-messageformat](https://www.npmjs.com/package/intl-messageformat) or [format-message](https://github.com/format-message/format-message), which do all the work in the browser. The approach taken by those libraries is more flexible as you can just load json files with translations in plain text and that's it, but it also means the library needs to ship a parser for the ICU message syntax, and it always has to have ship code for all the features that the ICU syntax supports, even features you might not use, making those libraries several times bigger.

The strategy used by this library assumes that your app will have some sort of built process that you can use to analyze and precompile
the messages in your app. This process spares the browser of the burden of doing the same process in the user's devices, resulting in
smaller and faster apps.

### Compile translations? WAT?

Ok, let's check it with an example.

Say you have a file with translations like this:
```js
{
  "plain": "Some text without interpolations",
  "interpolated": "A text where I interpolate {count} times",
  "time": "Now is {now, time}",
  "time-custom-format": "The hour is {now, time, hour}",
  "date": "Today is {today, date}",
  "date-custom-format": "Today is {today, date, abbr-full}",
  "number": "My favorite number is {n, number}",
  "percent": "My favorite number is {n, number, percent}",
  "pluralized": "I have {count, plural,=0 {no cats} =1 {one cat} other {{count} cats}}",
  "pluralized-with-hash": "I have {count, plural, zero {no cats} one {just # cat} other {# cats}}",
  "selected": "{gender, select, male {He is a good boy} female {She is a good girl} other {They are good fellas}}",
}
```

This babel plugin will precompile that file into something like this:
```js
import { __date, __interpolate, __number, __plural, __select, __time } from "precompile-intl-runtime";
export default {
  plain: "Some text without interpolations",
  interpolated: `A text where I interpolate ${__interpolate(count)} times`,
  time: now => `Now is ${__time(now)}`,
  "time-custom-format": now => `The hour is ${__time(now, "hour")}`,
  date: today => `Today is ${__date(today)}`,
  "date-custom-format": today => `Today is ${__date(today, "abbr-full")}`,
  number: n => `My favorite number is ${__number(n)}`,
  percent: n => `My favorite number is ${__number(n, "percent")}`,
  pluralized: count => `I have ${__plural(count, { 0: "no cats", 1: "one cat", h: `${__interpolate(count)} cats`})}`,
  "pluralized-with-hash": count => `I have ${__plural(count, { z: "no cats", o: `just ${count} cat`, h: `${count} cats`})}`,
  selected: gender => __select(gender, { male: "He is a good boy", female: "She is a good girl", other: "They are good fellas"})
}
```
As you see, translations with no logic remain unchanged, but translations that take arguments are transformed into functions that
take those arguments and use some runtime helpers from `precompile-intl-runtime`, the runtime counterpart of this library.

Although this output may seem very verbose, it is designed to minify extremely well.
Your minification step will transform keys like this:

```js
"pluralized-with-hash": "I have {count, plural, zero {no cats} one {just # cat} other {# cats}}",
--------------------------------------------------------------------------------------------------
"pluralized-with-hash":t=>`I have ${jt(t,{z:"no cats",o:`just ${t} cat`,h:`${t} cats`})}`
```

On average the minified output is between 5% and 10% smaller than the original input, and the runtime
library is much smaller than other alternatives and even smaller because it can be tree-shaken so only
the helpers your translations need are included, varying the weight of the library between 1k and just a few bytes.

If you store translations in `*.js` file, you may use template literals (strings with backticks). These strings may span multiple lines in the file.

```js
export default {
  notice: `
    Great to see you again!
    ... long text ...
  `
}
```

Note: you should not use JavaScript variables and expressions in these template literals. This syntax *is not* supported:

```js
const name = "Gregory"
export default {
  notice: `Hello, ${name}`
}
```

It will be better to use interpolations and pass values on call:

```js
// Translations file.
export default {
  notice: `Hello, {name}`
}

// Page, layout, etc. See `precompile-intl-runtime`.
$t('notice', {values: {name: 'Gregory'}})
```


Changelog.md
# 0.5.2
- Fix edge case when two plurals are contiguous with nothing in between.

# 0.4.0
- Support template literals in when translations are defined in javascript files (#17)
# 0.4.0-beta.2
- Initial support for `scale` in number skeletons.
# 0.4.0-beta.1
- Transform to ES Modules.
# 0.4.0-beta.0
- Support number skeletons. Support might not be complete yet.

test/index.test.js
import buildPlugin from '../dist';
import fs from 'fs';
import path from 'path';
import babel from '@babel/core';
const plugin = buildPlugin();

it("does not import functions if all keys are regular keys", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'plain-keys-only', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'plain-keys-only', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can transform interpolations", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-interpolations', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-interpolations', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can transforms `plural`", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-plurals', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-plurals', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can transforms plurals with # interpolations", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'plural-with-hash-interpolation', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'plural-with-hash-interpolation', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can transforms plurals with offsets", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'plural-with-offsets', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'plural-with-offsets', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can transforms `select`", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-select', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-select', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can format numbers", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'format-numbers', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'format-numbers', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can format dates", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'format-dates', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'format-dates', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can format times", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'format-times', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'format-times', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("can transforms nested plurals and selects", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'final-boss', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'final-boss', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("flattens the object, joining keys with a dot", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'flattening', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'flattening', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});

it("works with template literals", () => {
  let input = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-template-literals', 'input.js'), 'UTF8');
  let output = fs.readFileSync(path.join('test', 'fixtures', 'default', 'with-template-literals', 'output.js'), 'UTF8');
  const { code } = babel.transform(input, { plugins: [plugin] });
  expect(code).toBe(output);
});


test/fixtures/default/output.js
import { plural } from "helpers";
export default {
  nearby: "Find places near your location",
  kilometer: count => `${count} ${plural(count, {
    1: "kilometer",
    other: "kilometers"
  })}`
};

test/fixtures/default/html/input.js
export default {
  br: "Line with <br> line break",
  br2: "Line with <br> and interpolation {val}"
};


src/index.js
import { declare } from "@babel/helper-plugin-utils";
import { parse } from '@formatjs/icu-messageformat-parser';

const HELPERS_MAP = {
  1: "__interpolate",
  2: "__number",
  3: "__date",
  4: "__time",
  5: "__select",
  6: "__plural",
};
const PLURAL_ABBREVIATIONS = {
  zero: 'z',
  one: 'o',
  two: 't',
  few: 'f',
  many: 'm',
  other: 'h'
};

export default function build(runtimeImportPath = "precompile-intl-runtime") {
  return declare(({ types: t, assertVersion }, options) => {
    assertVersion("^7.0");
    let usedHelpers = new Set();
    let currentFunctionParams = new Set();
    let pluralsStack = [];

    function normalizePluralKey(key) {
      key = key.trim();
      let match = key.match(/^=(\d+)/);
      if (match) return parseInt(match[1], 10);
      return PLURAL_ABBREVIATIONS[key] || key;
    }

    function normalizeKey(key) {
      key = key.trim();
      let match = key.match(/^=(\d)/);
      if (match) return parseInt(match[1], 10);
      return key;
    }

    function buildCallExpression(entry) {
      let fnName = HELPERS_MAP[entry.type];
      usedHelpers.add(fnName === "__plural" && entry.offset !== 0 ? '__offsetPlural' : fnName);
      if (fnName === "__number") {
        return buildNumberCallExpression(entry);
      }
      if (fnName === "__interpolate" || fnName === "__date" || fnName === "__time") {
        let callArgs = [t.identifier(entry.value)];
        currentFunctionParams.add(entry.value);
        if (entry.style) callArgs.push(t.stringLiteral(entry.style));
        return t.callExpression(t.identifier(fnName), callArgs);
      }
      if (fnName === '__plural') {
        pluralsStack.push(entry)
      }
      let options = t.objectExpression(
        Object.keys(entry.options).map(key => {
          let objValueAST = entry.options[key].value;
          let objValue;
          if (objValueAST.length === 1 && objValueAST[0].type === 0) {
            objValue = t.stringLiteral(objValueAST[0].value);
          } else {
            objValue = objValueAST.length === 1 ? buildCallExpression(objValueAST[0]) : buildTemplateLiteral(objValueAST);
          }
          let normalizedKey = fnName === '__plural' ? normalizePluralKey(key) : normalizeKey(key);
          return t.objectProperty(
            typeof normalizedKey === "number"
              ? t.numericLiteral(normalizedKey)
              : t.identifier(normalizedKey),
            objValue
          );
        })
      );
      if (fnName === "__plural") {
        pluralsStack.pop();
      }
      currentFunctionParams.add(entry.value);
      let fnIdentifier = t.identifier(fnName);
      let callArguments = [t.identifier(entry.value)];
      if (fnName === "__plural" && entry.offset !== 0) {
        fnIdentifier =  t.identifier("__offsetPlural");
        callArguments.push(t.numericLiteral(entry.offset));
      }
      callArguments.push(options)
      return t.callExpression(fnIdentifier, callArguments);
    }

    function buildNumberCallExpression(entry) {
      let callArgs = [];
      if (entry.style?.parsedOptions?.scale) {
        // For number skeletons containing `scale/100` and similar, which is not an option I18n understand,
        // it replaces `__number(n, options) by `__number(n / 100, options)`.
        callArgs.push(
          t.binaryExpression("/", t.identifier(entry.value), t.numericLiteral(entry.style.parsedOptions.scale))
        );
        delete entry.style.parsedOptions.scale
      } else {
        callArgs.push(t.identifier(entry.value));
      }
      currentFunctionParams.add(entry.value);
      if (entry.style) {
        if (typeof entry.style === 'string') {
          // TODO: Special rule for 'percent' ?
          callArgs.push(t.stringLiteral(entry.style))
        } else if (Object.keys(entry.style.parsedOptions).length > 0) {
          // TODO: Being a compiler gives us the chance to be better at some things than other libraries
          // For instance, when using `{style: 'unit', unit: XXXXX}` unit can only be one of a known list of units
          // (see https://stackoverflow.com/questions/68035616/invalid-unit-argument-for-intl-numberformat-with-electric-units-volt-joule)
          // Any unit but those will throw a runtime error, but we could error or at least show a warning to the user
          // when invalid. P.e. a common mistake could be to use `unit: 'km'` (wrong) instead of the correct (`unit: 'kilometer'`).
          let options = t.objectExpression(
            Object.keys(entry.style.parsedOptions).map(key => {
              let val = entry.style.parsedOptions[key]
              return t.objectProperty(
                t.identifier(key),
                typeof val === "number" ? t.numericLiteral(val) : t.stringLiteral(val),
              );
            })
          );
          callArgs.push(options);
        }
      }
      return t.callExpression(t.identifier('__number'), callArgs);
    }

    function buildTemplateLiteral(ast) {
      let quasis = [];
      let expressions = [];
      for (let i = 0; i < ast.length; i++) {
        let entry = ast[i];
        switch(entry.type) {
          case 0: // literal
            quasis.push(
              t.templateElement(
                { value: entry.value, raw: entry.value },
                i === ast.length - 1 // tail
              )
            );
            break;
          case 1: // intepolation
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            if (i === 0) quasis.push(t.templateElement({ value: '', raw: '' }, false));
            break;
          case 2: // Number format
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            break;
          case 3: // Date format
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            break;
          case 4: // Time format
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            break;
          case 5: // select
            expressions.push(buildCallExpression(entry));
            break;
          case 6: // plural
            expressions.push(buildCallExpression(entry));
            break;
          case 7: // # interpolation
            let lastPlural = pluralsStack[pluralsStack.length - 1];
            if (lastPlural.offset !== null && lastPlural.offset !== 0) {
              expressions.push(
                t.binaryExpression(
                  "-",
                  t.identifier(lastPlural.value),
                  t.numericLiteral(lastPlural.offset)
                )
              );
            } else {
              expressions.push(t.identifier(lastPlural.value));
            }
            if (i === 0) quasis.push(t.templateElement({ value: '', raw: '' }, false));
            break;
          default:
            debugger;
        }
        if (i === ast.length - 1 && entry.type !== 0) {
          quasis.push(t.templateElement({ value: '', raw: '' }, true));
        }
      }
      // If the number of quasis must be one more than the number of expressions (because expressions go
      // in between). If that's not the case it means we need an empty string as first quasis.
      if (quasis.length === expressions.length) {
        quasis.unshift(t.templateElement({ value: '', raw: '' }, false));
      }
      return t.templateLiteral(quasis, expressions);
    }

    function buildFunction(ast) {
      currentFunctionParams = new Set();
      pluralsStack = [];
      let body = ast.length === 1 ? buildCallExpression(ast[0]) : buildTemplateLiteral(ast);
      return t.arrowFunctionExpression(
        Array.from(currentFunctionParams).sort().map(p => t.identifier(p)),
        body
      );
    }

    function flattenObjectProperties(object, propsArray, currentPrefix) {
      return object.properties.forEach(op => {
        let keyIsStringLiteral = t.isStringLiteral(op.key);
        let name = keyIsStringLiteral ? op.key.value : op.key.name;
        keyIsStringLiteral = keyIsStringLiteral || !!currentPrefix;
        name = currentPrefix ? currentPrefix + '.' + name : name;
        if (t.isObjectExpression(op.value)) {
          flattenObjectProperties(op.value, propsArray, name);
        } else {
          let key = keyIsStringLiteral ? t.stringLiteral(name) : t.identifier(name);
          propsArray.push([key, op.value]);
        }
      });
    }
    return {
      name: "precompile-intl",

      visitor: {
        Program: {
          enter() {
            usedHelpers = new Set();
          },
          exit(path) {
            if (usedHelpers.size > 0) {
              let importDeclaration = t.importDeclaration(
                Array.from(usedHelpers).sort().map(name => t.importSpecifier(t.identifier(name), t.identifier(name)))
              , t.stringLiteral(runtimeImportPath));
              path.unshiftContainer("body", importDeclaration);
            }
          }
        },
        ObjectProperty({ node }) {
          if (t.isStringLiteral(node.value)) {
            let icuAST = parse(node.value.value, {ignoreTag: true});
            if (icuAST.length === 1 && icuAST[0].type === 0) return;
            node.value = buildFunction(icuAST);
          } else if (t.isTemplateLiteral(node.value) && node.value.quasis.length == 1) {
            const quasis_value = node.value.quasis[0].value;
            let icuAST = parse(quasis_value.raw, {ignoreTag: true});
            if (icuAST.length === 1 && icuAST[0].type === 0) return;
            node.value = buildFunction(icuAST);
          }
        },
        VariableDeclarator({ node }) {
          if (t.isStringLiteral(node.init)) {
            let icuAST = parse(node.init.value);
            if (icuAST.length === 1 && icuAST[0].type === 0) return;
            node.init = buildFunction(icuAST);
          }
        },
        ExportDefaultDeclaration({ node }) {
          let properties = [];
          flattenObjectProperties(node.declaration, properties);
          node.declaration = t.objectExpression(
            properties.map(([k, v]) => {
              t.objectProperty(k,v)
              if (t.isIdentifier(k) && t.isIdentifier(v) && k.name === v.name) {
                return t.objectProperty(k, v, false, true);
              } else {
                return t.objectProperty(k,v)
              }
            })
          );
        }
      }
    };
  });
};


src/index.ts
import { declare } from "@babel/helper-plugin-utils";
import type { 
  Expression,
  Identifier,
  ObjectExpression,
  StringLiteral,
  TemplateElement
} from "@babel/types";
import { 
  isIdentifier,
  isObjectExpression,
  isObjectProperty,
  isStringLiteral,
} from "@babel/types";
import type { 
  ArgumentElement,
  DateElement,
  ExtendedNumberFormatOptions,
  MessageFormatElement,
  NumberElement,
  PluralElement,
  SelectElement,
  TimeElement,
  TYPE 
} from '@formatjs/icu-messageformat-parser';
import { parse } from '@formatjs/icu-messageformat-parser';
import { 
  isArgumentElement,
  isDateElement,
  isDateTimeSkeleton,
  isLiteralElement,
  isNumberElement,
  isNumberSkeleton,
  isPoundElement,
  isSelectElement,
  isTagElement,
  isTimeElement,
} from '@formatjs/icu-messageformat-parser/types.js';


type HelperFunctions = "__interpolate" | "__number" | "__date" | "__time" | "__select" | "__plural" | "__offsetPlural";
const HELPERS_MAP: Record<TYPE, HelperFunctions> = {
  0: "__interpolate", // This should not happen
  1: "__interpolate",
  2: "__number",
  3: "__date",
  4: "__time",
  5: "__select",
  6: "__plural",
  7: "__interpolate", // This should not happen
  8: "__interpolate", // This should not happen
};
type PluralTypes = 'zero' | 'one' | 'two' | 'few' | 'many' | 'other';
type PluralAbbreviation = 'z' | 'o' | 't' | 'f' | 'm' | 'h';
const PLURAL_ABBREVIATIONS: Record<PluralTypes, PluralAbbreviation> = {
  zero: 'z',
  one: 'o',
  two: 't',
  few: 'f',
  many: 'm',
  other: 'h'
};
type PropertyTuple = [Identifier | StringLiteral, any];

export default function build(runtimeImportPath = "precompile-intl-runtime") {
  return declare(({ types: t, assertVersion }, options) => {
    assertVersion("^7.0");
    let usedHelpers: Set<HelperFunctions> = new Set();
    let currentFunctionParams: Set<string> = new Set();
    let pluralsStack: PluralElement[] = [];

    function normalizePluralKey(key: string | PluralTypes): number | PluralAbbreviation {
      key = key.trim();
      let match = key.match(/^=(\d+)/);
      if (match) return parseInt(match[1], 10);
      return PLURAL_ABBREVIATIONS[key as PluralTypes] || key;
    }

    function normalizeKey(key: string): number | string {
      key = key.trim();
      let match = key.match(/^=(\d)/);
      if (match) return parseInt(match[1], 10);
      return key;
    }

    function buildCallExpression(entry: MessageFormatElement) {
      if (isLiteralElement(entry)) throw new Error('Literal elements not supported!');
      if (isTagElement(entry)) throw new Error('Tag elements not supported!');
      if (isPoundElement(entry)) throw new Error('Pound elements not supported!');
      if (isNumberElement(entry)) {
        return buildNumberCallExpression(entry);
      }
      if (isDateElement(entry) || isTimeElement(entry)) {
        return buildDateOrTimeCallExpression(entry)
      }
      if (isArgumentElement(entry)) {
        return buildInterpolateCallExpression(entry);
      }
      if (isSelectElement(entry)) {
        return buildSelectCallExpression(entry);
      }
      return buildPluralCallExpression(entry);
    }

    function buildNumberCallExpression(entry: NumberElement) {
      usedHelpers.add(HELPERS_MAP[entry.type]);
      let callArgs = [];
      if (isNumberSkeleton(entry.style) && entry.style.parsedOptions.scale) {
        callArgs.push(
          t.binaryExpression("/", t.identifier(entry.value), t.numericLiteral(entry.style.parsedOptions.scale))
        );
        delete entry.style.parsedOptions.scale        
      } else {
        callArgs.push(t.identifier(entry.value));
      }
      currentFunctionParams.add(entry.value);
      if (isNumberSkeleton(entry.style)) {
        if (Object.keys(entry.style.parsedOptions).length > 0) {
          // TODO: Being a compiler gives us the chance to be better at some things than other libraries
          // For instance, when using `{style: 'unit', unit: XXXXX}` unit can only be one of a known list of units
          // (see https://stackoverflow.com/questions/68035616/invalid-unit-argument-for-intl-numberformat-with-electric-units-volt-joule)
          // Any unit but those will throw a runtime error, but we could error or at least show a warning to the user
          // when invalid. P.e. a common mistake could be to use `unit: 'km'` (wrong) instead of the correct (`unit: 'kilometer'`).
          let keys = Object.keys(entry.style.parsedOptions) as (keyof ExtendedNumberFormatOptions)[];
          let options = t.objectExpression(
            keys.map(key => {
              if (!isNumberSkeleton(entry.style)) throw new Error('The entry should have had a number skeleton')
              let val = entry.style.parsedOptions[key] as number | string;
              return t.objectProperty(
                t.identifier(key),
                typeof val === "number" ? t.numericLiteral(val) : t.stringLiteral(val),
              );
            })
          );
          callArgs.push(options);
        }
      } else if (typeof entry.style === 'string') {
        callArgs.push(t.stringLiteral(entry.style));
      }
      return t.callExpression(t.identifier('__number'), callArgs);
    }

    function buildDateOrTimeCallExpression(entry: DateElement | TimeElement) {
      let fnName = HELPERS_MAP[entry.type];
      usedHelpers.add(fnName);
      let callArgs: (Identifier | StringLiteral)[]  = [t.identifier(entry.value)];
      currentFunctionParams.add(entry.value);
      if (isDateTimeSkeleton(entry.style)) throw new Error('Datetime skeletons not supported yet');
      if (entry.style) {
        callArgs.push(t.stringLiteral(entry.style));
      }
      return t.callExpression(t.identifier(fnName), callArgs);      
    }

    function buildInterpolateCallExpression(entry: ArgumentElement) {
      let fnName = HELPERS_MAP[entry.type];
      usedHelpers.add(fnName);
      currentFunctionParams.add(entry.value);
      return t.callExpression(t.identifier(fnName), [t.identifier(entry.value)]);

    }

    function buildPluralCallExpression(entry: PluralElement) {
      let fnName = HELPERS_MAP[entry.type];
      pluralsStack.push(entry)
      usedHelpers.add(entry.offset !== 0 ? '__offsetPlural' : '__plural');
      let options = t.objectExpression(
        Object.keys(entry.options).map(key => {
          let objValueAST = entry.options[key].value;
          let objValue: Expression;
          if (objValueAST.length === 1 && objValueAST[0].type === 0) {
            objValue = t.stringLiteral(objValueAST[0].value);
          } else {
            objValue = objValueAST.length === 1 ? buildCallExpression(objValueAST[0]) : buildTemplateLiteral(objValueAST);
          }
          let normalizedKey = normalizePluralKey(key);
          return t.objectProperty(
            typeof normalizedKey === "number"
              ? t.numericLiteral(normalizedKey)
              : t.identifier(normalizedKey),
            objValue
          );
        })
      );
      pluralsStack.pop();
      currentFunctionParams.add(entry.value);
      if (entry.offset !== 0) {
        return t.callExpression(t.identifier("__offsetPlural"), [t.identifier(entry.value), t.numericLiteral(entry.offset), options]);
      } else {
        return t.callExpression(t.identifier(fnName), [t.identifier(entry.value), options]);
      }
    }

    function buildSelectCallExpression(entry: SelectElement) {
      let fnName = HELPERS_MAP[entry.type];
      usedHelpers.add(fnName);
      entry.options
      let options = t.objectExpression(
        Object.keys(entry.options).map(key => {
          let objValueAST = entry.options[key].value;
          let objValue: Expression;
          if (objValueAST.length === 1 && objValueAST[0].type === 0) {
            objValue = t.stringLiteral(objValueAST[0].value);
          } else {
            objValue = objValueAST.length === 1 ? buildCallExpression(objValueAST[0]) : buildTemplateLiteral(objValueAST);
          }
          let normalizedKey = normalizeKey(key);
          return t.objectProperty(
            typeof normalizedKey === "number" ? t.numericLiteral(normalizedKey) : t.identifier(normalizedKey),
            objValue
          );
        })
      );
      currentFunctionParams.add(entry.value);
      return t.callExpression(t.identifier(fnName), [t.identifier(entry.value), options]);
    }

    function buildTemplateLiteral(ast: MessageFormatElement[]) {
      let quasis: TemplateElement[] = [];
      let expressions: Expression[] = [];
      for (let i = 0; i < ast.length; i++) {
        let entry = ast[i];
        switch(entry.type) {
          case 0: // literal
            quasis.push(
              t.templateElement(
                // { value: entry.value, raw: entry.value }, this is not valid anymore?
                { cooked: entry.value, raw: entry.value },
                i === ast.length - 1 // tail
              )
            );
            break;
          case 1: // intepolation
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            if (i === 0) quasis.push(t.templateElement({ cooked: '', raw: '' }, false));
            break;
          case 2: // Number format
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            break;
          case 3: // Date format
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            break;
          case 4: // Time format
            expressions.push(buildCallExpression(entry));
            currentFunctionParams.add(entry.value);
            break;
          case 5: // select
            expressions.push(buildCallExpression(entry));
            break;
          case 6: // plural
            expressions.push(buildCallExpression(entry));
            break;
          case 7: // # interpolation
            let lastPlural = pluralsStack[pluralsStack.length - 1];
            if (lastPlural.offset !== null && lastPlural.offset !== 0) {
              expressions.push(
                t.binaryExpression(
                  "-",
                  t.identifier(lastPlural.value),
                  t.numericLiteral(lastPlural.offset)
                )
              );
            } else {
              expressions.push(t.identifier(lastPlural.value));
            }
            if (i === 0) quasis.push(t.templateElement({ cooked: '', raw: '' }, false));
            break;
          default:
            debugger;
        }
        if (i === ast.length - 1 && entry.type !== 0) {
          quasis.push(t.templateElement({ cooked: '', raw: '' }, true));
        }
      }

      if (quasis.length === expressions.length && quasis.length === 0) {
        return t.stringLiteral(''); // If there's no data, return an empty string. No need to use backquotes.
      }
      // If the number of quasis must be one more than the number of expressions (because expressions go
      // in between). If that's not the case it means we need an empty string as first quasis.
      while (quasis.length <= expressions.length) {
        quasis.unshift(t.templateElement({ cooked: '', raw: '' }, false));
      }

      return t.templateLiteral(quasis, expressions);
    }

    function buildFunction(ast: MessageFormatElement[]) {
      currentFunctionParams = new Set();
      pluralsStack = [];
      let body = ast.length === 1 ? buildCallExpression(ast[0]) : buildTemplateLiteral(ast);
      if (Array.from(currentFunctionParams).length === 0) {
        return body;
      }
      return t.arrowFunctionExpression(
        Array.from(currentFunctionParams).sort().map(p => t.identifier(p)),
        body
      );
    }

    function flattenObjectProperties(object: ObjectExpression, propsArray: PropertyTuple[], currentPrefix?: string) {
      return object.properties.forEach(op => {
        if (!isObjectProperty(op)) throw new Error('Exported objects can only contain regular properties');
        if (!isStringLiteral(op.key) && !isIdentifier(op.key)) throw new Error('Object keys for translations can only contain strings or identifiers');
        let name = isStringLiteral(op.key) ? op.key.value : op.key.name;
        name = currentPrefix ? currentPrefix + '.' + name : name;
        if (t.isObjectExpression(op.value)) {
          flattenObjectProperties(op.value, propsArray, name);
        } else {
          let key = (isStringLiteral(op.key) || !!currentPrefix) ? t.stringLiteral(name) : t.identifier(name);
          propsArray.push([key, op.value]);
        }
      });
    }
    return {
      name: "precompile-intl",

      visitor: {
        Program: {
          enter() {
            usedHelpers = new Set();
          },
          exit(path) {
            if (usedHelpers.size > 0) {
              let importDeclaration = t.importDeclaration(
                Array.from(usedHelpers).sort().map(name => t.importSpecifier(t.identifier(name), t.identifier(name)))
              , t.stringLiteral(runtimeImportPath));
              path.unshiftContainer("body", importDeclaration);
            }
          }
        },
        ObjectProperty({ node }) {
          if (t.isStringLiteral(node.value)) {
            let icuAST = parse(node.value.value, {ignoreTag: true});
            if (icuAST.length === 1 && icuAST[0].type === 0) return;
            node.value = buildFunction(icuAST);
          } else if (t.isTemplateLiteral(node.value) && node.value.quasis.length == 1) {
            const quasis_value = node.value.quasis[0].value;
            let icuAST = parse(quasis_value.raw, {ignoreTag: true});
            if (icuAST.length === 1 && icuAST[0].type === 0) return;
            node.value = buildFunction(icuAST);
          }
        },
        VariableDeclarator({ node }) {
          if (t.isStringLiteral(node.init)) {
            let icuAST = parse(node.init.value);
            if (icuAST.length === 1 && icuAST[0].type === 0) return;
            node.init = buildFunction(icuAST);
          }
        },
        ExportDefaultDeclaration({ node }) {
          // let properties: (Identifier | StringLiteral)[] = [];
          let properties: PropertyTuple[] = [];
          if (!isObjectExpression(node.declaration)) throw new Error('The default export must be an object');
          flattenObjectProperties(node.declaration, properties);
          node.declaration = t.objectExpression(
            properties.map(([k, v]) => {
              t.objectProperty(k,v)
              if (t.isIdentifier(k) && t.isIdentifier(v) && k.name === v.name) {
                return t.objectProperty(k, v, false, true);
              } else {
                return t.objectProperty(k,v)
              }
            })
          );
        }
      }
    };
  });
};
